### PR TITLE
Source statistics stream

### PR DESCRIPTION
**Pull request by blootsvoets**
- Update to Confluent 5.0.0 / Kafka 2.0.0
- Use a Kafka Processor API for SourceStatistics
- Manage streams from a single `StreamMaster`
- Make stream definitions inside the `StreamWorker`, not in a `StreamGroup`
- specify stream properties per stream
- use `java.time.Duration` for intervals
- Made code ready for converting the `Monitor` implementations to `StreamWorker`.

### CODE REVIEW

File: src/main/java/org/radarcns/config/SourceStatisticsMonitorConfig.java [hunk 1]
```diff
@@ -2,8 +2,9 @@
 
 import com.fasterxml.jackson.annotation.JsonProperty;
 import java.util.List;
+import org.radarcns.stream.statistics.SourceStatisticsStream;
 
-public class SourceStatisticsMonitorConfig {
+public class SourceStatisticsMonitorConfig extends SingleStreamConfig {
```
Comment (yatharthranjan) [inline src/main/java/org/radarcns/config/SourceStatisticsMonitorConfig.java]: Can be renamed to `SourceStatisticsStreamConfig`

File: src/main/java/org/radarcns/stream/KafkaStreamFactory.java [hunk 2]
```diff
@@ -16,62 +16,63 @@
...
-            try {
-                masters.add((StreamMaster) Class.forName(streamType).newInstance());
-            } catch (InstantiationException | IllegalAccessException | ClassNotFoundException e) {
-                logger.error("Cannot instantiate StreamMaster class {}", streamType, e);
-            } catch (ClassCastException e) {
-                logger.error("Given type {} is not a StreamMaster", streamType);
-            }
-        }
+    private StreamMaster master(Stream<? extends SingleStreamConfig> configs) {
+        return new StreamMaster(config, configs);
```
Comment (yatharthranjan) [inline src/main/java/org/radarcns/stream/KafkaStreamFactory.java]: The instance variable can be renamed to radarProperties or something similar for clarity.

File: src/main/java/org/radarcns/stream/empatica/E4HeartRateStream.java [hunk 3]
```diff
@@ -18,29 +18,27 @@
...
-    public E4HeartRateStream(Collection<StreamDefinition> definitions, int numThread,
-            StreamMaster master, RadarPropertyHandler properties) {
-        super(definitions, numThread, master, properties, logger);
+public class E4HeartRateStream extends
+        SensorStreamWorker<ObservationKey, EmpaticaE4InterBeatInterval> {
+    @Override
+    protected void initialize() {
+        defineWindowedSensorStream(
+                "android_empatica_e4_inter_beat_interval",
+                "android_empatica_e4_heart_rate");
```
Comment (yatharthranjan) [inline src/main/java/org/radarcns/stream/empatica/E4HeartRateStream.java]: Is `"android_empatica_e4_inter_beat_interval"` supposed to be here ? Also `SensorStreamWorker<ObservationKey, EmpaticaE4InterBeatInterval>` should be `SensorStreamWorker<ObservationKey, EmpaticaE4HeartRate>` ?

File: src/main/java/org/radarcns/stream/phone/PhoneUsageAggregationStream.java [hunk 4]
```diff
@@ -1,42 +1,42 @@
...
-            StreamMaster master, RadarPropertyHandler properties) {
-        super(definitions, numThread, master, properties, logger);
+public class PhoneUsageAggregationStream extends
+        SensorStreamWorker<ObservationKey, PhoneUsageEvent> {
+    @Override
+    protected void initialize() {
+        defineStream(
+                "android_phone_usage_event_output",
+                "android_phone_usage_event_aggregated",
+                Duration.ofDays(1));
```
Comment (yatharthranjan) [inline src/main/java/org/radarcns/stream/phone/PhoneUsageAggregationStream.java]: These should also be windowed as neither the Rest API nor the dashboard has a way of consuming these.

File: src/main/java/org/radarcns/stream/phone/PhoneUsageStream.java [hunk 5]
```diff
@@ -39,12 +37,16 @@
...
-    public PhoneUsageStream(Collection<StreamDefinition> definitions, int numThread,
-            StreamMaster master, RadarPropertyHandler properties) {
-        super(definitions, numThread, master, properties, logger);
+    public PhoneUsageStream() {
         playStoreLookup = new PlayStoreLookup(CACHE_TIMEOUT, MAX_CACHE_SIZE);
     }
 
+    @Override
+    protected void initialize() {
+        defineSensorStream("android_phone_usage_event");
```
Comment (yatharthranjan) [inline src/main/java/org/radarcns/stream/phone/PhoneUsageStream.java]: This should be windowed as well

File: src/main/java/org/radarcns/stream/statistics/SourceStatisticsStream.java [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+    }
+
+    protected Duration getInterval() {
+        return interval;
+    }
+
+    @SuppressWarnings("PMD.OptimizableToArrayCall")
+    private Topology getTopology() {
+        Topology builder = new Topology();
+        GenericAvroDeserializer genericReader = new GenericAvroDeserializer();
```
Comment (yatharthranjan) [inline src/main/java/org/radarcns/stream/statistics/SourceStatisticsStream.java]: What happens here if the schema id of a payload does not exist in the schema registry ?

File: src/main/java/org/radarcns/stream/empatica/E4HeartRateStream.java (hunk 3 above)
Comment (blootsvoets) [reply src/main/java/org/radarcns/stream/empatica/E4HeartRateStream.java]: So what it does is consume `InterBeatInterval` and dynamically compute `HeartRate`. That also means that the `EmpaticaE4InterBeatInterval` generics is as it should be. I'll add a comment in the superclass about the generic parameters.

Comment (yatharthranjan) [reply src/main/java/org/radarcns/stream/empatica/E4HeartRateStream.java]: Sorry my bad. I just noticed that emaptica does not give heart rate directly and it has to be calculated from interbeat interval. Please ignore

File: src/main/java/org/radarcns/stream/phone/PhoneUsageAggregationStream.java (hunk 4 above)
Comment (blootsvoets) [reply src/main/java/org/radarcns/stream/phone/PhoneUsageAggregationStream.java]: I don't see a way to aggregate them though, or how to show them in the Dashboard at all. If there is a suggestion about this, we can add a new implementation for this. These streams are primarily meant to make the output files in HDFS more manageable.

File: src/main/java/org/radarcns/stream/phone/PhoneUsageStream.java (hunk 5 above)
Comment (blootsvoets) [reply src/main/java/org/radarcns/stream/phone/PhoneUsageStream.java]: See above, this is only meant to improve the output in HDFS, not for the dashboard.

File: src/main/java/org/radarcns/stream/phone/PhoneUsageAggregationStream.java (hunk 4 above)
Comment (yatharthranjan) [reply src/main/java/org/radarcns/stream/phone/PhoneUsageAggregationStream.java]: But the output topics from this stream are never used in HDFS connector topics config. So thats why i was wondering where is this used.

File: src/main/java/org/radarcns/stream/statistics/SourceStatisticsStream.java (hunk 6 above)
Comment (blootsvoets) [reply src/main/java/org/radarcns/stream/statistics/SourceStatisticsStream.java]: Updated default exception handler in KafkaProperty with `props.put(StreamsConfig.DEFAULT_DESERIALIZATION_EXCEPTION_HANDLER_CLASS_CONFIG,
                LogAndContinueExceptionHandler.class.getName());`

File: src/main/java/org/radarcns/stream/phone/PhoneUsageAggregationStream.java (hunk 4 above)
Comment (blootsvoets) [reply src/main/java/org/radarcns/stream/phone/PhoneUsageAggregationStream.java]: Good point, that is an error, it should definitely be added to the config.

File: src/main/java/org/radarcns/stream/statistics/SourceStatisticsStream.java (hunk 6 above)
Comment (yatharthranjan) [reply src/main/java/org/radarcns/stream/statistics/SourceStatisticsStream.java]: Great. This also helps in other cases as mentioned in #71

File: src/main/java/org/radarcns/stream/phone/PhoneUsageAggregationStream.java (hunk 4 above)
Comment (blootsvoets) [reply src/main/java/org/radarcns/stream/phone/PhoneUsageAggregationStream.java]: See https://github.com/RADAR-base/RADAR-Docker/pull/157.

**Pull request merged**